vcd/resource_vcd_catalog_item.go

```diff
@@ -293,14 +305,37 @@ func createOrUpdateCatalogItemMetadata(d *schema.ResourceData, meta interface{})
 		for _, k := range toBeRemovedMetadata {
 			err := vAppTemplate.DeleteMetadataEntry(k)
 			if err != nil {
-				return fmt.Errorf("error deleting metadata: %s", err)
+				return fmt.Errorf("error deleting metadata from catalog item's associated vApp template: %s", err)
 			}
 		}
 		// Add new metadata
 		for k, v := range newMetadata {
 			err := vAppTemplate.AddMetadataEntry(types.MetadataStringValue, k, v.(string))
 			if err != nil {
-				return fmt.Errorf("error adding metadata: %s", err)
+				return fmt.Errorf("error adding metadata to catalog item's associated vApp template: %s", err)
+			}
+		}
+	}
+	if d.HasChange("catalog_item_metadata") {
+		oldRaw, newRaw := d.GetChange("catalog_item_metadata")
```

Comment:
Let's add TODO here or even if you have time please try: this code could be moved to function with generics :) and replaced through all the code. The difference in these code manipulation is an error message and parent object(like `catalogItem`) to manipulate with meta data

Reply:
I was implementing that suggestion about generics, but maybe it complicates too much this PR. Is it OK if I leave the TODO and I do the generics refactor in a separated PR?

Reply:
Ok. Yep, this can be quite a big change through many resources. So we should mark it or if you want it can be your next PR.

Reply:
Just for traceability, leaving the PR for that here: #853